### PR TITLE
PROC-1219: Fix rule return result check so that it supports method calls

### PR DESCRIPTION
**Pull request by robertdanci**

### CODE REVIEW

File: proctor-tomcat-deps-provided/pom.xml [hunk 1]
```diff
@@ -32,14 +32,6 @@
                     <groupId>org.apache.tomcat</groupId>
                     <artifactId>tomcat-jasper-el</artifactId>
                 </exclusion>
-                <exclusion>
```
Comment (robertdanci) [inline proctor-tomcat-deps-provided/pom.xml:35]: The `proctor-tomcat-deps` does not bring in `tomcat-jsp-api` and neither `tomcat-servlet-api`.

File: proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java [hunk 2]
```diff
@@ -85,11 +83,24 @@ public void testNonBooleanShouldFail() {
                     .hasMessageContaining("Received non-boolean return value");
         }
         {
-            // numeric result becomes String, not boolean
-            assertThatThrownBy(() -> ruleEvaluator.evaluateBooleanRule("${'tr'}${'ue'}", emptyMap()))
-                    .isInstanceOf(IllegalArgumentException.class)
-                    .hasMessageContaining("Received non-boolean return value");
+            // String 'true' is true
+            assertTrue(ruleEvaluator.evaluateBooleanRule("${'tr'}${'ue'}", emptyMap()));
```
Comment (zacharygoodwin) [inline proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java:87]: I do not think this is a valid format, ${'true'} would be valid but I do not think this format should be supported. cc @stevenchen-indeed for confirmation

Comment (stevenchen-indeed) [reply proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java:87]: Yes this does not look like convention. If it does not error out, maybe it is a corner case that was not caught before,

Comment (vanhaxl) [reply proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java:87]: IMO, it's okay as I think no one writes the rule like ${'true'}. But this will help validate the custom function like author's team need: placementFilter.matchesMetadata(INDEED_PLUS)

Comment (zacharygoodwin) [reply proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java:87]: Looking into this we already allow it to get inputted into a test this just makes parsing it valid.

File: proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java [hunk 3]
```diff
@@ -85,11 +83,24 @@ public void testNonBooleanShouldFail() {
...
+
+    @Test
+    public void testStandaloneMethodCalls() {
+        final Map<String, Object> context = ImmutableMap.of("context", new Temp());
+        for (String rule : ImmutableList.of("${context.isValid()}", "${context.isFortyTwo('42')}")) {
+            assertTrue("rule '" + rule + "' should be true for " + context, ruleEvaluator.evaluateBooleanRule(rule, context));
+        }
+
+        for (String rule : ImmutableList.of("${!context.isValid()}", "${context.isFortyTwo('47')}")) {
+            assertFalse("rule '" + rule + "' should be true for " + context, ruleEvaluator.evaluateBooleanRule(rule, context));
```
Comment (zacharygoodwin) [inline proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java:99]: Nitpick: "' should be false for "

**Review by zacharygoodwin — APPROVED** [review]

File: proctor-common/src/main/java/com/indeed/proctor/common/RuleEvaluator.java [hunk 4]
```diff
@@ -131,51 +131,26 @@ public boolean evaluateBooleanRule(final String rule, @Nonnull final Map<String,
         }
 
         final ELContext elContext = createElContext(values);
-        final ValueExpression ve = expressionFactory.createValueExpression(elContext, rule, boolean.class);
```
[the remark this answers is not in the document]
Comment (robertdanci) [reply proctor-common/src/main/java/com/indeed/proctor/common/RuleEvaluator.java:134]: The reason is that we still check the return type (see line 136) and that method now expects a String.
If we are happy with removing the call to `checkRuleIsBooleanType` we can coerce to boolean instead. 

Tbh I don't think there's much value of checking that type is boolean _at evaluation time_ since we already did that when the rule was added on the UI (in `RuleVerifyUtils`). Was thinking to remove it but we have a test that checks this (TestRuleEvaluator#testNonBooleanShouldFail) and wanted to minimize the changes but happy to do it if we have consensus.

File: proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java (hunk 2 above)
Comment (robertdanci) [reply proctor-common/src/test/java/com/indeed/proctor/common/TestRuleEvaluator.java:87]: Yeah, there was already a test with this rule (which probably should not be allowed on the UI) so I just 'fixed' it here in light of the new approach.

File: proctor-common/src/main/java/com/indeed/proctor/common/RuleEvaluator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stevenchen-indeed) [reply proctor-common/src/main/java/com/indeed/proctor/common/RuleEvaluator.java:134]: Thanks Robert.
What i found is, if we keep checkRuleIsBooleanType the same, changing to string "final ValueExpression ve = expressionFactory.createValueExpression(elContext, rule, String.class);" will still result in property not found in checkRuleIsBooleanType.
javax.el.PropertyNotFoundException: Property [isValid] not found on type [com.indeed.proctor.common.TestRuleEvaluator$Temp]
There i am not sure what's the difference between String and boolean in this case.

At least that's what i saw when i run from unit tests. Do you see different behavior? 

Or do you mean Class<?> type = ve.getType(elContext); may not be required, using string and check "true" or "false" is just as reliable?

[the remark this answers is not in the document]
Comment (stevenchen-indeed) [reply proctor-common/src/main/java/com/indeed/proctor/common/RuleEvaluator.java:134]: To be honest i am also trying to learn how does the proctor api validate the rule.
Taking your unit tests for example,

final Map<String, Object> context = ImmutableMap.of("context", new Temp());

The Temp can be any class that you need in your application. How does Proctor Api get access to those custom class?
